### ISSUE TITLE
Stop PG-JDBC threads

### ISSUE THREAD

**Issue by tlf30** [impossibl/pgjdbc-ng#487]
Hello @kdubb ,
I am probably missing something here, but how do I 'stop' ng-pgjdbc? I do not have any connectrions open to the database, but I have several PG-JDBC I/O threads running. Is there a way to stop them, in the sense of shutting down jdbc, or disconnecting it from the database?

```
Thread-1             	 RUNNABLE 	 5 	 Normal
globalEventExecutor-1-11 	 TIMED_WAITING 	 5 	 Normal
Finalizer            	 WAITING 	 8 	 Daemon
PG-JDBC I/O (1)      	 RUNNABLE 	 5 	 Daemon
Thread-2             	 TIMED_WAITING 	 5 	 Normal
PG-JDBC I/O (3)      	 RUNNABLE 	 5 	 Daemon
PG-JDBC I/O (2)      	 RUNNABLE 	 5 	 Daemon
PG-JDBC I/O (1)      	 TIMED_WAITING 	 5 	 Daemon
PG-JDBC I/O (3)      	 TIMED_WAITING 	 5 	 Daemon
PG-JDBC I/O (3)      	 RUNNABLE 	 5 	 Daemon
PG-JDBC I/O (2)      	 TIMED_WAITING 	 5 	 Daemon
PG-JDBC I/O (2)      	 RUNNABLE 	 5 	 Daemon
Common-Cleaner       	 TIMED_WAITING 	 8 	 Daemon
DestroyJavaVM        	 RUNNABLE 	 5 	 Normal
Reference Handler    	 RUNNABLE 	 10 	 Daemon
PG-JDBC I/O (1)      	 RUNNABLE 	 5 	 Daemon
PG-JDBC I/O (2)      	 RUNNABLE 	 5 	 Daemon
Attach Listener      	 RUNNABLE 	 5 	 Daemon
PG-JDBC I/O (3)      	 RUNNABLE 	 5 	 Daemon
Signal Dispatcher    	 RUNNABLE 	 9 	 Daemon
PG-JDBC I/O (1)      	 RUNNABLE 	 5 	 Daemon

```

Thanks,
Trevor

**Comment by kdubb** [comment]
After closing all the connections you can use
```java
PGDriver.cleanup()
```
to shut down the Netty I/O

**Comment by tlf30** [comment]
Perfect, thank you @kdubb. 
Can this be added to the documentation?

**Comment by tlf30** [comment]
I am getting a NPE when attempting to use this:
```
java.lang.NullPointerException
	at com.impossibl.postgres.protocol.v30.ServerConnectionShared.shutdown(ServerConnectionShared.java:112)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.HashMap$ValueSpliterator.forEachRemaining(HashMap.java:1675)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at com.impossibl.postgres.protocol.v30.ServerConnectionShared.waitForShutdown(ServerConnectionShared.java:122)
	at com.impossibl.postgres.jdbc.PGDriver.cleanup(PGDriver.java:178)
	at io.tlf.outside.server.database.Database.unload(Database.java:165)
```

**Comment by kdubb** [comment]
It seems as if shutdown is being called twice and, after looking at the code, I can see that it's not guarded against.

I can work on a fix for double shutdown but it would be great if you could verify that is the case by guarding against it yourself.

**Comment by tlf30** [comment]
I just verified by using two approaches, first using the classic system.out.println to say how many times I am unloading. Then to be safe I moved it to the last line in my program (just before the quit command terminates the jvm, so it is not possible to run more than once). 

```
    public void run(ConsoleProvider provider, String... args) {
        cm.getServer().quit();
        provider.close();
        PGDriver.cleanup();
        System.exit(0);
    }
```
The only usage of `cleanup()` is once in my program, and once in ng-jdbc:
![image](https://user-images.githubusercontent.com/10089434/80985471-bdce7580-8ddb-11ea-96f1-ada1c6e73040.png)

```
java.lang.NullPointerException
	at com.impossibl.postgres.protocol.v30.ServerConnectionShared.shutdown(ServerConnectionShared.java:112)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.HashMap$ValueSpliterator.forEachRemaining(HashMap.java:1675)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at com.impossibl.postgres.protocol.v30.ServerConnectionShared.waitForShutdown(ServerConnectionShared.java:122)
	at com.impossibl.postgres.jdbc.PGDriver.cleanup(PGDriver.java:178)
	at com.impossibl.postgres.jdbc.PGDriver$cleanup.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:130)
	at io.tlf.outside.server.command.QuitCommand.run(QuitCommand.groovy:36)
	at io.tlf.outside.server.console.ConsoleManager.exec(ConsoleManager.java:115)
	at io.tlf.outside.server.console.ConsoleManager.exec(ConsoleManager.java:101)
	at io.tlf.outside.server.console.ConsoleManager.cmd(ConsoleManager.java:75)
	at io.tlf.outside.server.console.ConsoleManager$1.run(ConsoleManager.java:53)
```

 EDIT: I will start working on a standalone test case

**Comment by kdubb** [comment]
That seems odd. The only way for `eventLoopGroup` to become `null` is by calling `shutdown`.

You can also try patching `ServerConnectionShared.java` by replacing `shutdown` with the following:

```java

  private Future<?> shutdown() {

    if (eventLoopGroup == null) {
      return ImmediateEventExecutor.INSTANCE.newSucceededFuture(null);
    }

    Future<?> res = eventLoopGroup.shutdownGracefully(10, 100, TimeUnit.MILLISECONDS);
    eventLoopGroup = null;
    return res;
  }
```

**Comment by tlf30** [comment]
I was able to recreate the issue in a standalone test:
https://github.com/tlf30/pgjdbc-ng-test4

**Comment by tlf30** [comment]
OK, I have made some progress on this. It only fails if there are no open connections to the database. 
If I open a connection, then call `PGDriver.cleanup();` everything works as expected. This can be done in the test case by commenting out the `con.close();` at the end before the `cleanup`

I have updated the test to print the running threads at the end.

**Comment by kdubb** [comment]
Yeah that makes sense... it's essentially the same as a double close (i.e. `eventLoopGroup` is null).

I made a bit of a mistake in my hectic Monday morning and I meant to create a PR with the fix but I accidentally pushed it to develop 🤷‍♂️... so, _if_ that commit passes CI checks it will be pushed to `0.9-SNAPSHOT`. It's essentially the code I posted above.

**Comment by kdubb** [comment]
@tlf30 This is fixed in the latest `0.9-SNAPSHOT`. Verified with your test case. Thanks!

**Comment by tlf30** [comment]
Awesome, thank you @kdubb! 
I really appreciate your time on fixing this issues.
